nbqa/save_source.py

```diff
@@ -19,6 +20,73 @@
 IGNORE_LINE_REPLACEMENT = "pass  # nbqa"
 
 
+def _is_src_code_indentation_valid(source: str) -> bool:
+    """
+    Return True is the indentation of the input source code is valid.
+
+    Parameters
+    ----------
+    source : str
+        Source code present in the notebook cell
+
+    Returns
+    -------
+    bool
+        True if the source indentation is valid otherwise False
+    """
+    valid: bool = True
+    try:
+        ast.parse(source)
+    except IndentationError:
+        valid = False
+    except SyntaxError:
+        # Ignore any other exceptions. Let the syntax issues
+        # be reported by the third party tool.
+        pass
+
+    return valid
+
+
+def _handle_line_magic_indentation(
+    source: List[str], line_magic: str
+) -> Tuple[str, str]:
+    """
+    Handle the indentation of line magics. Remove unnecessary indentation.
+
+    Parameters
+    ----------
+    source : List[str]
+        Source code of the notebook cell
+    line_magic : str
+        Line magic present in the notebook cell
+
+    Returns
+    -------
+    Tuple[str, str]
+        Tuple of replacement line and the original line magic statement.
+    """
+    leading_space = "".join(takewhile(lambda c: c == " ", line_magic))
+    token = secrets.token_hex(3)
+
+    # preserve the leading spaces and check if the syntax is valid
+    line_tokenised = f"{leading_space}{IGNORE_LINE_REPLACEMENT}{token}"
+
+    if leading_space and not _is_src_code_indentation_valid(
+        "".join(source + [line_tokenised])
+    ):
+        # Remove the line magic indentation assuming these leading spaces
+        # lead the source to have invalid indentation
+        # Currently we don't check if the original source
+        # code itself had IndentationError
```

Comment:
That's a good point...if we're going to do this, reckon there would be much overhead in checking that the source code _without_ this extra line has no indentation error? Perhaps we need an extra test-case in which the indentation error doesn't come from the magic?

Reply:
On a second thought, what you're doing is probably alright - if the original source code itself has an indentation error, then any tool which uses ast.parse (e.g. `black`) will report an error anyway